kairos-server/src/l1_sync/event_manager.rs

```diff
@@ -0,0 +1,85 @@
+use std::sync::Arc;
+
+use casper_event_toolkit::fetcher::{Fetcher, Schemas};
+use casper_event_toolkit::metadata::CesMetadataRef;
+use casper_event_toolkit::rpc::client::CasperClient;
+
+use crate::state::transactions::{Deposit, Signed, Transaction};
+use crate::state::BatchStateManager;
+
+use super::error::L1SyncError;
+
+pub struct EventManager {
+    next_event_id: u32,
+    fetcher: Option<Fetcher>,
+    schemas: Option<Schemas>,
+    batch_service: Arc<BatchStateManager>,
+}
+
+impl EventManager {
+    pub fn new(batch_service: Arc<BatchStateManager>) -> Self {
+        EventManager {
+            next_event_id: 0,
+            fetcher: None,
+            schemas: None,
+            batch_service,
+        }
+    }
+
+    /// Initializes state by building CES fetcher and obtaining schemas.
+    pub async fn initialize(
+        &mut self,
+        rpc_url: &str,
+        contract_hash: &str,
+    ) -> Result<(), L1SyncError> {
+        tracing::info!("Initializing event manager");
+
+        let client = CasperClient::new(rpc_url);
+        let metadata = CesMetadataRef::fetch_metadata(&client, contract_hash).await?;
+        tracing::debug!("Metadata fetched successfully");
+
+        let fetcher = Fetcher {
+            client,
+            ces_metadata: metadata,
+        };
+        let schemas = fetcher.fetch_schema().await?;
+        tracing::debug!("Schemas fetched successfully");
+
+        self.fetcher = Some(fetcher);
+        self.schemas = Some(schemas);
+
+        Ok(())
+    }
```

Comment:
> The logic from `initialize` can go into `new` the mutex.

I don't want initialization logic in the constructor, as it should be straightforward and non-blocking.

> The `EventManager` service should start initialized, it should not be a command.

I initially made it a command, because I wanted to run initialization in the background, without blocking server start. After thinking it out again I see that we should synchronize state, before processing requests.

Changes done in https://github.com/cspr-rad/kairos/pull/96/commits/1b6cae2cd5610e853be8ed1439568ddcfb7eaa95.

Reply:
I'd still like to see `initialize` merged into `new`.
But doing that in an eventual followup is fine, so I'll approve.

Reply:
Fixed in https://github.com/cspr-rad/kairos/pull/96/commits/50a3654c33e72267c89df53df61ca72a2bcca9eb.